libs/server/Custom/CustomCommandManagerSession.cs

```diff
@@ -13,53 +14,81 @@ internal sealed class CustomCommandManagerSession
         readonly CustomCommandManager customCommandManager;
 
         // These session specific arrays are indexed by the same ID as the arrays in CustomCommandManager
-        readonly (CustomTransactionProcedure, int)[] sessionTransactionProcMap;
-        readonly CustomProcedure[] sessionCustomProcMap;
-
+        ExpandableMap<CustomTransactionProcedureWithArity> sessionTransactionProcMap;
+        ExpandableMap<CustomProcedure> sessionCustomProcMap;
 
         public CustomCommandManagerSession(CustomCommandManager customCommandManager)
         {
             this.customCommandManager = customCommandManager;
-            sessionTransactionProcMap = new (CustomTransactionProcedure, int)[CustomCommandManager.MaxRegistrations];
-            sessionCustomProcMap = new CustomProcedure[CustomCommandManager.MaxRegistrations];
+            sessionTransactionProcMap = new ExpandableMap<CustomTransactionProcedureWithArity>(CustomCommandManager.MinMapSize, 0, byte.MaxValue);
```

Comment:
how does this get updated when procedures are added to the main structure? see how this is handled in `main` - we might need a similar approach.

Reply:
The session map is a "cache" of the main map, if the id is not in the session map then it checks the main map and updates the session map.